keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,857 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding "predicate" expressions to admission webhooks, as an extension to the
+existing `rules` to define the scope of a webhook. A `predicate` is a
+[CEL](https://github.com/google/cel-spec) expression that must evaluate to true for the admission
+request to be sent to the webhook. If a `predicate` evaluates to false, the webhook is skipped for
+that request (implicitly allowed).
+
+<<[UNRESOLVED naming ]>>
+I'm not satisfied with the name `predicates` for this feature. Other ideas considered:
+- filters, prefilters
+- match{Predicates, Filters, Criteria, Expressions, Constraints, ...}
+<<[/UNRESOLVED]>>
+
+## Motivation
+
+**Reliability:** Admission webhooks continue to be an operational sore spot for many Kubernetes
+users. Webhooks that target cluster critical resources put the admission controller backing the
+webhook in the critical path of cluster stability. Even if tools like namespace scoping are used to
+avoid circular-dependencies and exclude critical system resources, a webhook outage can still have a
+major impact on cluster availability. This proposal aims to mitigate (but not eliminate) these
+issues by allowing webhooks to be more narrowly scoped and targeted.
+
+**Performance:** Admission webhooks sit in the critical request path for write-requests. Validating
+webhooks can be run in parallel, but Mutating webhooks must be run in serial (up to 2 times!). This
+makes webhooks extremely latency sensitive, and even a webhook that doesn't do any work still needs
+to pay the network round-trip cost.
+
+**Supportability:** For hosted or managed Kubernetes distributions, webhooks can be a problem when
+they interfere with requests by managed components. The existing criteria for filtering out requests
+are insufficient for many use cases, and aren't easily appended with provider rules.
+
+_What about [CEL for Admission Control](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control)?_
+`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
+for admission webhooks, but it is intentionally not attempting to cover every possible use case.
+This proposal aims to improve the situation for those webhooks that cannot be migrated.
+
+### User Stories
+
```

Comment:
The webhook implementation of Pod Security Admission is a good example here I think: you might prefer to be able to run a healthy control plane in `kube-system` even if the otherwise-mandatory PSA webhook wasn't healthy.

Reply:
You can do that today with a namespace selector though. In fact, since PSA operates on namespace labels, you could exclude any namespace labeled as `privileged`.

I chose the use cases here specifically because they're things that don't currently have a good solution.